cylc/flow/data_store_mgr.py

```diff
@@ -925,63 +894,76 @@ def update_task_proxies(self, updated_tasks=None):
+        self.updated_state_families.clear()
+        while self.state_update_families:
+            for fam_id in self.state_update_families:
+                self._family_ascent_point_update(fam_id)
+                break
+
+    def _family_ascent_point_update(self, fp_id):
+        fp_added = self.added[FAMILY_PROXIES]
+        fp_data = self.data[self.workflow_id][FAMILY_PROXIES]
+        # TODO: deprecate try-except with event driven delta collection
+        #       or n-distance data window.
+        try:
+            if fp_id in fp_data:
+                fam_node = fp_data[fp_id]
+            else:
+                fam_node = fp_added[fp_id]
+        except KeyError:
+            self.updated_state_families.add(fp_id)
+            self.state_update_families.remove(fp_id)
             return
-
-        for point_string in cycle_points:
-            # For each cycle point, construct a family state tree
-            # based on the first-parent single-inheritance tree
-            c_task_states = self.delta_cycle_states.get(point_string, None)
-            if c_task_states is None:
-                continue
-            c_fam_task_states = {}
-            c_fam_task_is_held = {}
-
-            for key in c_task_states:
-                state, is_held = c_task_states[key]
-                if state is None:
-                    continue
-
-                for parent in self.ancestors.get(key, []):
-                    if parent == key:
-                        continue
-                    c_fam_task_states.setdefault(parent, set())
-                    c_fam_task_states[parent].add(state)
-                    c_fam_task_is_held.setdefault(parent, False)
-                    if is_held:
-                        c_fam_task_is_held[parent] = is_held
-
-            for fam, child_states in c_fam_task_states.items():
-                state = extract_group_state(child_states)
-                fp_id = (
-                    f'{self.workflow_id}{ID_DELIM}'
-                    f'{point_string}{ID_DELIM}{fam}')
-                if state is None or (
-                        fp_id not in family_proxies and
-                        fp_id not in self.added[FAMILY_PROXIES]):
-                    continue
-                # Since two fields strings are reassigned,
-                # it should be safe without copy.
-                fp_delta = PbFamilyProxy(
-                    id=fp_id,
-                    stamp=f'{fp_id}@{time()}',
-                    state=state,
-                    is_held=c_fam_task_is_held[fam]
-                )
-                self.updated[FAMILY_PROXIES].setdefault(
-                    fp_id, PbFamilyProxy()).MergeFrom(fp_delta)
+        child_fam_nodes = [
+            n_id
+            for n_id in fam_node.child_families
+            if n_id not in self.updated_state_families
+        ]
+        for child_fam_id in child_fam_nodes:
+            self._family_ascent_point_update(child_fam_id)
```

Comment:
This recursively works its way to the bottom of the tree then works its way up to `fp_id`?

Would it be possible to do this the other way around, starting at the task which has changed state and bubbling that update up the family hierarchy?

Reply:
> This recursively works its way to the bottom of the tree then works its way up to fp_id?

Not quite, it starts from whatever first parent family task(s) flag to be updated..

> Would it be possible to do this the other way around, starting at the task which has changed state and bubbling that update up the family hierarchy?

Yes, that's exactly how it works. The updated tasks add their first parent to `self.state_update_families`, however, if it's a middle family then you need to check if any child families are also in `self.state_update_families`  (that's the recursive bit)... This Means all the families are checked (which is a super fast set lookup), but only those that actually need updating are changed (deltas created).

Reply:
Ah, heck, I don't quite understand that? Might need an example.

> Not quite, it starts from whatever first parent family task(s) flag to be updated..

What does "first parent family task" mean?

> if it's a middle family then you need to check if any child families are also in `self.state_update_families`

I think I'm misunderstanding something here. For the following inheritance hierarchy:

* root
  * FOO
    * foot
    * BAR
        * bart

I would expect that:

* If `foot` changed `FOO` should be recomputed.
* If `bar` changed `FOO` and `BAR` should be recomputed.

In this example is `FOO` a middle family?

Reply:
Ah, ok, I think I've been thick, the `child_fam_nodes ` are taken from `self.updated_state_families` so this is just an efficiency improvement to avoid recomputing the family tree twice?

Reply:
Sorry "first-parent-family that task(s)"

Reply:
> the `child_fam_nodes` are taken from `self.updated_state_families`

The `child_fam_nodes` are taken from `self.state_update_families`  nodes (`PbFamilyProxy.child_families`), and filtered to not include those already updated (`self.updated_state_families`). The filtering avoids re-computation, but I could probably broaden the filtered ones to any family that has been checked (not just the updated ones).

Reply:
> but I could probably broaden the filtered ones to any family that has been checked (not just the updated ones).

Actually I'm not sure it's possible to check the same family twice as is.. So no need to widen net, if a child family is checked but not updated it's parent (or ancestor) would have been checked (added to `self.updated_state_families`), so it won't be hit again.

Reply:
Ok, I think I have a rough idea what it's doing now, thanks for clearing it up.